src/librustc_privacy/lib.rs

```diff
@@ -172,168 +172,162 @@ struct EmbargoVisitor<'a, 'tcx: 'a> {
+                let (public_ty, exported_ty) = self.is_public_exported_ty(&ty);
+
+                for impl_item in impl_items {
+                    if impl_item.vis == hir::Public {
+                        if public_ty {
+                            self.public_items.insert(impl_item.id);
+                        }
+                        if exported_ty {
+                            self.exported_items.insert(impl_item.id);
                         }
                     }
-                    _ => true,
-                };
-                let tr = self.tcx.impl_trait_ref(self.tcx.map.local_def_id(item.id));
-                let public_trait = tr.clone().map_or(false, |tr| {
-                    if let Some(node_id) = self.tcx.map.as_local_node_id(tr.def_id) {
-                        self.exported_items.contains(&node_id)
-                    } else {
-                        true
-                    }
-                });
+                }
+            }
 
-                if public_ty || public_trait {
-                    for impl_item in impl_items {
-                        match impl_item.node {
-                            hir::ConstImplItem(..) => {
-                                if (public_ty && impl_item.vis == hir::Public)
-                                    || tr.is_some() {
-                                    self.exported_items.insert(impl_item.id);
-                                }
-                            }
-                            hir::MethodImplItem(ref sig, _) => {
-                                let meth_public = match sig.explicit_self.node {
-                                    hir::SelfStatic => public_ty,
-                                    _ => true,
-                                } && impl_item.vis == hir::Public;
-                                if meth_public || tr.is_some() {
-                                    self.exported_items.insert(impl_item.id);
-                                }
-                            }
-                            hir::TypeImplItem(_) => {}
-                        }
+            // It's not known until monomorphization if a trait impl item should be reachable
+            // from external crates or not. So, we conservatively mark all of them exported and
+            // the reachability pass (middle::reachable) marks all exported items as reachable.
+            // For example of private trait impl for private type that shoud be reachable see
+            // src/test/auxiliary/issue-11225-3.rs
+            hir::ItemImpl(_, _, _, Some(ref trait_ref), ref ty, ref impl_items) => {
+                let (public_ty, _exported_ty) = self.is_public_exported_ty(&ty);
+                let (public_trait, _exported_trait) = self.is_public_exported_trait(trait_ref);
+
+                if public_ty && public_trait {
+                    self.public_items.insert(item.id);
+                }
+                self.exported_items.insert(item.id);
```

Comment:
Er actually, this seems like it may just want to be guarded by `exported_trait` as if the trait isn't exported then there's no need to export the impl even if the type is exported (no one can reach it unless through wonky reasons `middle::reachable` should handle)